File: levelledmobs-plugin-bukkit/src/main/java/io/github/arcaneplugins/levelledmobs/bukkit/util/PlayerUtils.java
package io.github.arcaneplugins.levelledmobs.bukkit.util;

import java.util.function.Predicate;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.PlayerInventory;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class PlayerUtils {

    private PlayerUtils() throws IllegalAccessException {
        throw new IllegalAccessException("Illegal instantiation of utility class");
    }


    public static @Nullable FoundItemInHandResult findItemStackInEitherHand(
        final @NotNull Player player,
        final @NotNull Predicate<@Nullable ItemStack> predicate
    ) {
        final PlayerInventory inventory = player.getInventory();

        if(predicate.test(inventory.getItemInMainHand())) {
            return new FoundItemInHandResult(inventory.getItemInMainHand(), true);
        } else if(predicate.test(inventory.getItemInOffHand())) {
            return new FoundItemInHandResult(inventory.getItemInOffHand(), false);
        }

        return null;
    }

    /**
     * Object returned by {@link PlayerUtils#findItemStackInEitherHand(Player, Predicate)}.
     *
     * @param itemStack  The {@link ItemStack} that was found in the search.
     * @param inMainHand {@code true} if the item was found in the main hand, {@code false} if
     *                   the item was found in the off-hand.
     * @since 4.0.0
     */
    public record FoundItemInHandResult(
        @NotNull ItemStack itemStack,
        boolean inMainHand
    )  {}

}

File: levelledmobs-plugin-bukkit/src/main/kotlin/io/github/arcaneplugins/levelledmobs/bukkit/util/PlayerUtils.kt
package io.github.arcaneplugins.levelledmobs.bukkit.util

import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import java.util.function.Predicate

object PlayerUtils {
    fun findItemStackInEitherHand(
        player: Player,
        predicate: Predicate<ItemStack?>
    ): FoundItemInHandResult?{
        val inventory = player.inventory

        if (predicate.test(inventory.itemInMainHand)) {
            return FoundItemInHandResult(inventory.itemInMainHand, true)
        } else if (predicate.test(inventory.itemInOffHand)) {
            return FoundItemInHandResult(inventory.itemInOffHand, false)
        }

        return null
    }

    class FoundItemInHandResult(
        val itemStack: ItemStack,
        val inMainHand: Boolean
    )
}